File: _dist_/bin/UrlTools.js
import { printf, toConsole, interpolate, toText, isNullOrWhiteSpace, split } from "./.fable/fable-library.3.2.9/String.js";
import { toString } from "./.fable/fable-library.3.2.9/Types.js";
import { decompressFromEncodedURIComponent, compressToEncodedURIComponent } from "../../_snowpack/pkg/lz-string.js";
import { length } from "./.fable/fable-library.3.2.9/Seq.js";
import { map, choose, tryHead } from "./.fable/fable-library.3.2.9/Array.js";
import { fromString } from "./.fable/Thoth.Json.5.1.0/Decode.fs.js";

function setGetParam(encodedJson) {
    if (!(((arg00) => {
        history.pushState(arg00);
    }) == null)) {
        const hashPieces = split(window.location.hash, ["?"], null, 1);
        const hash = ((!(hashPieces == null)) ? (!isNullOrWhiteSpace(hashPieces[0])) : false) ? hashPieces[0] : "";
        const params = new URLSearchParams();
        params.set("data", encodedJson);
        const newUrl = toText(interpolate("%P()//%P()%P()%P()?%P()", [window.location.protocol, window.location.host, window.location.pathname, hash, toString(params)]));
        history.pushState({
            path: newUrl,
        }, "", newUrl);
    }
}

const encodeUrl = compressToEncodedURIComponent;

const decodeUrl = decompressFromEncodedURIComponent;

const URLSearchParamsExist = 'URLSearchParams' in window;

export function updateUrlBy(mapFn) {
    if (URLSearchParamsExist) {
        const hashPieces = window.location.hash.split("?");
        const params = new URLSearchParams(hashPieces[1]);
        const newUrl = toText(interpolate("%P()//%P()%P()%P()?%P()", [window.location.protocol, window.location.host, window.location.pathname, mapFn((window.location.hash == null) ? "" : window.location.hash).split("?")[0], toString(params)]));
        history.pushState({
            path: newUrl,
        }, "", newUrl);
    }
}

export function updateUrlWithData(json) {
    setGetParam(encodeUrl(json));
}

function $007CKeyValuesFromHash$007C_$007C(hash) {
    if (isNullOrWhiteSpace(hash)) {
        return void 0;
    }
    else {
        const search = hash.split("?");
        if (length(search) > 1) {
            return tryHead(choose((tupledArg) => {
                if (tupledArg[0] === "data") {
                    return tupledArg[1];
                }
                else {
                    return void 0;
                }
            }, map((kv) => [kv.split("=")[0], kv.split("=")[1]], search[1].split("\u0026"))));
        }
        else {
            return void 0;
        }
    }
}

export function restoreModelFromUrl(decoder, defaultValue) {
    const matchValue = window.location.hash;
    const activePatternResult12427 = $007CKeyValuesFromHash$007C_$007C(matchValue);
    if (activePatternResult12427 != null) {
        const v = activePatternResult12427;
        const modelResult = fromString(decoder, decodeUrl(decodeURIComponent(v)));
        if (modelResult.tag === 1) {
            toConsole(printf("%A"))(modelResult.fields[0]);
            return defaultValue;
        }
        else {
            return modelResult.fields[0];
        }
    }
    else {
        return defaultValue;
    }
}

File: src/client/fsharp/UrlTools.fs
module FantomasTools.Client.UrlTools

open Fable.Core.JsInterop
open Fable.Core
open Thoth.Json
open Browser.Types
open Browser
open System

let private setGetParam (encodedJson: string) : unit =
    if not (isNullOrUndefined history.pushState) then
        let hashPieces =
            window.location.hash.Split([| '?' |], StringSplitOptions.RemoveEmptyEntries)

        let hash =
            if
                not (isNullOrUndefined hashPieces)
                && not (String.IsNullOrWhiteSpace(hashPieces.[0]))
            then
                hashPieces.[0]
            else
                ""

        let ``params`` = URLSearchParams.Create("")
        ``params``.set ("data", encodedJson)

        let newUrl =
            $"%s{window.location.protocol}//%s{window.location.host}%s{window.location.pathname}%s{hash}?%s{``params``.ToString()}"

        let currentUrl = window.location.toString ()

        if currentUrl <> newUrl then
            history.pushState ({| path = newUrl |}, "", newUrl)

let private encodeUrl (_x: string) : string =
    import "compressToEncodedURIComponent" "lz-string"

let private decodeUrl (_x: string) : string =
    import "decompressFromEncodedURIComponent" "lz-string"

/// A link carries the whole model, compressed, in its `data` parameter. Every link ever shared is
/// read by the pair below, so they are kept apart from the window: what they do can be checked
/// against links people posted years ago without a browser in the room.
let encodeData (json: string) : string = encodeUrl json

let decodeData (data: string) : string = JS.decodeURIComponent data |> decodeUrl

/// Called rather than bound, so that reading this module does not need a window.
let private urlSearchParamsExist () : bool = emitJsExpr () "'URLSearchParams' in window"

let updateUrlBy (mapFn: string -> string) : unit =
    if urlSearchParamsExist () then
        let hashPieces = window.location.hash.Split('?')
        let ``params`` = URLSearchParams.Create(hashPieces.[1])

        let safeHash =
            if isNullOrUndefined window.location.hash then
                ""
            else
                window.location.hash

        let newHash = (mapFn safeHash).Split('?').[0]

        let newUrl =
            $"%s{window.location.protocol}//%s{window.location.host}%s{window.location.pathname}%s{newHash}?%s{``params``.ToString()}"

        history.pushState ({| path = newUrl |}, "", newUrl)

let updateUrlWithData json = setGetParam (encodeUrl json)

/// The `data` parameter of a hash, if it has one.
let tryDataFromHash (hash: string) : string voption =
    if String.IsNullOrWhiteSpace(hash) then
        ValueNone
    else
        let search = hash.Split('?')

        if Array.length search > 1 then
            search.[1].Split('&')
            |> Array.map (fun kv -> kv.Split('=').[0], kv.Split('=').[1])
            |> Array.tryPick (fun (k, v) -> if k = "data" then Some v else None)
            |> ValueOption.ofOption
        else
            ValueNone

/// Reads a model out of a hash. A link that cannot be read falls back to the default rather than
/// throwing: the hash is whatever was pasted into the address bar.
let restoreModelFromHash (hash: string) decoder defaultValue =
    match tryDataFromHash hash with
    | ValueNone -> defaultValue
    | ValueSome data ->
        match Decode.fromString decoder (decodeData data) with
        | Result.Ok m -> m
        | Error err ->
            printfn "%A" err
            defaultValue

let restoreModelFromUrl decoder defaultValue =
    restoreModelFromHash Browser.Dom.window.location.hash decoder defaultValue
